### PR TITLE
[API][Promotion] Index coupons only for requested promotion

### PR DESCRIPTION
**Pull request by Zales0123**
| Q               | A
| --------------- | ---
| Bug fix?        | yes |
| New feature?    | no |
| BC breaks?      | yes |
| Related tickets | fixes https://github.com/Sylius/Sylius/issues/8025, closes https://github.com/Sylius/Sylius/pull/8024 |
| License         | MIT |

Thanks @alcaeus for pointing this out.

### CODE REVIEW

**Comment by lchrusciel** [comment]
Thanks @alcaeus and @Zales0123! Nice job 👍

File: src/Sylius/Component/Promotion/Repository/PromotionCouponRepositoryInterface.php [hunk 1]
```diff
@@ -43,4 +43,11 @@ public function countByCodeLength(int $codeLength): int;
      * @return PromotionCouponInterface|null
      */
     public function findOneByCodeAndPromotionCode(string $code, string $promotionCode): ?PromotionCouponInterface;
+
+    /**
+     * @param string $promotionCode
+     *
+     * @return iterable
+     */
+    public function findByPromotionCode(string $promotionCode): iterable;
```
Comment (pamil) [inline src/Sylius/Component/Promotion/Repository/PromotionCouponRepositoryInterface.php]: All find* repository methods returns array of all results, this one returns a paginator, we should mention it somewhere or change the method name at all (`createPaginatorForPromotion`?).

Comment (pamil) [inline src/Sylius/Component/Promotion/Repository/PromotionCouponRepositoryInterface.php]: Anyway, it's a BC break and should be mentioned in` UPGRADE.md`.

Comment (alcaeus) [reply src/Sylius/Component/Promotion/Repository/PromotionCouponRepositoryInterface.php]: Does it need to return a paginator? IIRC the paginator is applied automatically if a route is paginated?

Comment (pamil) [reply src/Sylius/Component/Promotion/Repository/PromotionCouponRepositoryInterface.php]: It works only when there's no repository method defined (here is) and for the criteria specified in routing. I'm not sure whether we can query by a nested field `promotion.code` using it.

Comment (alcaeus) [reply src/Sylius/Component/Promotion/Repository/PromotionCouponRepositoryInterface.php]: Ah, your right. In that case, I'd suggest putting paginator in the name 👍

Comment (pamil) [reply src/Sylius/Component/Promotion/Repository/PromotionCouponRepositoryInterface.php]: Btw. supporting relations in `criteria` looks like a useful feature.

Comment (lchrusciel) [reply src/Sylius/Component/Promotion/Repository/PromotionCouponRepositoryInterface.php]: good catch, thanks @pamil

**Pull request merged**